### DISCUSSION TITLE
RTK Query: Do we have a manual fetch function that follows cache rules as auto-fetch do

### DISCUSSION THREAD

**Discussion by xiaoyangfeng** [reduxjs/redux-toolkit#1776] [Q&A] [answered]
I have check the refetch function and trigger function from useQuery and useLazyQuery.
Looks like refetch will send a request to server and trigger will use the cache or send a request based on preferCacheValue.
However, what I am looking for is a function which applies same fetch rules as auto-fetch but triggered manually. For example, if I set refetchOnMountOrArgChange: 30, and the second call is fired after 30 secs, it should send a request otherwise return the cache.

**Comment by markerikson** [comment]
@xiaoyangfeng I'm not sure I understand what you're asking for here. Can you give a more detailed example?

**Reply by xiaoyangfeng** [reply]
phryneas had the answer.  And I would like to explain more.
My use case is a search function. URL would be /search/keyword, after enter or click search button, useNavigate will be used to change URL.
And I used useQuery in my code. 
`import { useSearchByPhraseQuery } from "./app/apis/search";
  const { data, isFetching } = useSearchByPhraseQuery(body);
`
I would like to send a new request to server if the search is 30 secs later.  It works fine if I keep changging keyword.
For example, /search/a and /search/b in turns. It would send new requests for a and b after 30 secs.
However, it does not work for same keyword, if the URL is /search/a, and keep clicking the search button. Not new request even after 30 secs.
That is why I would like to a manual trigger (follow the cache rules), then I could put it on my click handler and call it when the keyword is same.
Hope this could make my question more clear. And let me know if the sturcture I am using for this search function is not the best practice and I would not face this kind of issue if the best practise was applied.

**Reply by phryneas** [reply]
But when the user clicks, a request should definitely be made (the user is quite explicit about that), so why do the cache lookup in that case? Just call `refetch`.

**Reply by xiaoyangfeng** [reply]
> But when the user clicks, a request should definitely be made (the user is quite explicit about that), so why do the cache lookup in that case? Just call `refetch`.

What if the data oweners know the fetched data wont change too often and also want to have some limitations in front end.

**Reply by phryneas** [reply]
What hinders your user from repeatedly hammering F5 from getting that result? Usually, when a user already sees data, but wants it refetched, I'd personally say honor that wish.

**Reply by xiaoyangfeng** [reply]
lol. ok

**Reply by zousandian** [reply]
> What hinders your user from repeatedly hammering F5 from getting that result? Usually, when a user already sees data, but wants it refetched, I'd personally say honor that wish.

I have a similar need.
For example, when an `<SearchInput/>` has auto suggestion feature, just like Google, the search function is triggered by the onChange event. A user may edit their input, from `Rect` to `React` for example, which may trigger two times `Re` search. 
We can cache the `Re` result for a moment to improve the speed. But after later, when the user search something like `Reselect`, which trigger `Re` search again, we still want to make a request to get new data.

**Reply by phryneas** [reply]
I would assume that if you query to `"https://example.com/search?q=Re"` you would expect the result to be the same every time? On which grounds would you expect a different result between the first time, the second time and the third time you call it? 
From what I see the server has no additional data and will not be able to destinguish between three times the same request?

**Reply by zousandian** [reply]
> d will not be able to destinguish between three times the same request?

When query something like `/business/receipts/list?sn=202201...`, may new data be generated from other place every few moment.

**Reply by phryneas** [reply]
So this is a time-related thing? In that case, why not just reduce the cache-lifetime if something is not on screen for that endpoint? [`keepUnusedDataFor​`](https://redux-toolkit.js.org/rtk-query/api/createApi#keepunuseddatafor)

**Comment by phryneas** [comment, marked as answer]
You can do that by doing
```js
const request = dispatch(
    api.endpoints.myEndpoint.initiate(arg, {
        forceRefetch: 30,
    })
)
// make sure to call this before your component unmounts or the data will never be removed from the cache
request.unsubscribe()
```

**Reply by xiaoyangfeng** [reply]
Thanks, that works! Any plan to add this as a build in function to the result of useQuery and useLazyQuery?

**Reply by phryneas** [reply]
No, you are the first person to ever request such a functionality and it is pretty different from how the library is usually being used, so you'll have to stick to using it directly.
